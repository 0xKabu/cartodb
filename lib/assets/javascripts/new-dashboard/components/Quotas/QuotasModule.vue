<template>
  <ul class="grid">
    <li class="vertical-space grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile">
      <QuotaContainer>
        <QuotaWidget :name="$t(`Limits.storage`)" :quotaType="$t(`Limits.quota`)" :usedQuota="divideBaseTwo(usedStorage, getBaseTwo)" :availableQuota="divideBaseTwo(availableStorage, getBaseTwo)" :unit="getUnitFromBaseTwo(getBaseTwo)"></QuotaWidget>
      </QuotaContainer>
    </li>
    <li class="vertical-space grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile">
      <QuotaContainer>
        <QuotaWidget :name="$t(`Limits.geocoding`)" :quotaType="$t(`Limits.credits`)" :usedQuota="geocodingUsed" :availableQuota="geocodingAvailable"></QuotaWidget>
      </QuotaContainer>
    </li>
    <li class="vertical-space grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile">
      <QuotaContainer>
        <QuotaWidget :name="$t(`Limits.isolines`)" :quotaType="$t(`Limits.credits`)" :usedQuota="isolinesUsed" :availableQuota="isolinesAvailable" mode="compact"></QuotaWidget>
        <QuotaWidget :name="$t(`Limits.routing`)" :quotaType="$t(`Limits.credits`)" :usedQuota="routingUsed" :availableQuota="routingAvailable" mode="compact"></QuotaWidget>
      </QuotaContainer>
    </li>
  </ul>
</template>

<script>
import { mapState } from 'vuex';
import QuotaWidget from 'new-dashboard/components/Quotas/QuotaWidget';
import QuotaContainer from 'new-dashboard/components/Quotas/QuotaContainer';

export default {
  name: 'QuotasModule',
  components: {
    QuotaWidget,
    QuotaContainer
  },
  computed: {
    ...mapState({
      availableStorage: state => state.user.quota_in_bytes,
      remainingStorage: state => state.user.remaining_byte_quota,
      geocodingUsed: state => state.user.geocoding.monthly_use,
      geocodingAvailable: state => state.user.geocoding.quota,
      routingUsed: state => state.user.mapzen_routing.monthly_use,
      routingAvailable: state => state.user.mapzen_routing.quota ? state.user.mapzen_routing.quota : 0,
      isolinesUsed: state => state.user.here_isolines.monthly_use,
      isolinesAvailable: state => state.user.here_isolines.quota
    }),
    usedStorage () {
      return this.availableStorage - this.remainingStorage;
    },
    getBaseTwo () {
      return this.baseTwoRepresentation(this.availableStorage);
    }
  },
  methods: {
    baseTwoRepresentation (initialByte) {
      let i = 0;
      let value = initialByte;
      while (value > 1024) {
        value = value / 1024;
        i = i + 1;
      }
      return (i * 10);
    },
    getUnitFromBaseTwo (baseTwo) {
      if (baseTwo < 10) {
        return 'B';
      } else if (baseTwo < 20) {
        return 'Kb';
      } else if (baseTwo < 30) {
        return 'MB';
      } else if (baseTwo < 40) {
        return 'GB';
      } else if (baseTwo < 50) {
        return 'TB';
      } else if (baseTwo < 60) {
        return 'PB';
      } else if (baseTwo < 70) {
        return 'EB';
      } else {
        return '?';
      }
    },
    divideBaseTwo (number, baseTwo) {
      return number / Math.pow(2, baseTwo);
    }
  }
};
</script>

<style scoped lang="scss">
@import 'stylesheets/new-dashboard/variables';

.vertical-space {
  margin-bottom: 10px;
}

</style>
