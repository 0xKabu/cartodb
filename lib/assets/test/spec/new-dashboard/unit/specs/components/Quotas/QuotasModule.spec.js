import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import usersArray from '../../fixtures/users';
import QuotasModule from 'new-dashboard/components/Quotas/QuotasModule';

const localVue = createLocalVue();
localVue.use(Vuex);

let store = new Vuex.Store({
  state: {
    user: usersArray[2]
  }
});

const $t = key => key;

describe('QuotasModule.vue', () => {
  it('should render correct contents', () => {
    const quotasModule = shallowMount(QuotasModule, {
      store,
      localVue,
      mocks: {
        $t
      }
    });

    expect(quotasModule).toMatchSnapshot();
  });

  describe('Computed', () => {
    it('should return all the correct computed variables', () => {
      const quotasModule = shallowMount(QuotasModule, {
        store,
        localVue,
        mocks: {
          $t
        }
      });

      let availableStoarage = quotasModule.vm.availableStorage;
      let remainingStorage = quotasModule.vm.remainingStorage;
      let geocodingUsed = quotasModule.vm.geocodingUsed;
      let geocodingAvailable = quotasModule.vm.geocodingAvailable;
      let routingUsed = quotasModule.vm.routingUsed;
      let routingAvailable = quotasModule.vm.routingAvailable;
      let isolinesUsed = quotasModule.vm.isolinesUsed;
      let isolinesAvailable = quotasModule.vm.isolinesAvailable;
      let usedStorage = quotasModule.vm.usedStorage;
      let getBaseTwo = quotasModule.vm.getBaseTwo;

      expect(availableStoarage).toBe(1073741824);
      expect(remainingStorage).toBe(1066631168);
      expect(geocodingUsed).toBe(103);
      expect(geocodingAvailable).toBe(1200);
      expect(routingUsed).toBe(54);
      expect(routingAvailable).toBe(102);
      expect(isolinesUsed).toBe(134);
      expect(isolinesAvailable).toBe(1332);
      expect(usedStorage).toBe(7110656);
      expect(getBaseTwo).toBe(20);
    });

    it('should return correct unit if you give a base two', () => {
      const quotasModule = shallowMount(QuotasModule, {
        store,
        localVue,
        mocks: {
          $t
        }
      });

      let result1 = quotasModule.vm.getUnitFromBaseTwo(0);
      let result2 = quotasModule.vm.getUnitFromBaseTwo(12);
      let result3 = quotasModule.vm.getUnitFromBaseTwo(20);
      let result4 = quotasModule.vm.getUnitFromBaseTwo(35);
      let result5 = quotasModule.vm.getUnitFromBaseTwo(49);
      let result6 = quotasModule.vm.getUnitFromBaseTwo(51);
      let result7 = quotasModule.vm.getUnitFromBaseTwo(62);
      let result8 = quotasModule.vm.getUnitFromBaseTwo(70);

      expect(result1).toBe('B');
      expect(result2).toBe('Kb');
      expect(result3).toBe('MB');
      expect(result4).toBe('GB');
      expect(result5).toBe('TB');
      expect(result6).toBe('PB');
      expect(result7).toBe('EB');
      expect(result8).toBe('?');
    });

    it('should divide the number passed by the base two provided', () => {
      const quotasModule = shallowMount(QuotasModule, {
        store,
        localVue,
        mocks: {
          $t
        }
      });

      let result1 = quotasModule.vm.divideBaseTwo(0, 0);
      let result2 = quotasModule.vm.divideBaseTwo(10, 0);
      let result3 = quotasModule.vm.divideBaseTwo(2097152, 20);
      let result4 = quotasModule.vm.divideBaseTwo(132070244352, 30);

      expect(result1).toBe(0);
      expect(result2).toBe(10);
      expect(result3).toBe(2);
      expect(result4).toBe(123);
    });
  });

  describe('Methods', () => {
    it('should return correct base two of the number (ten by ten)', () => {
      const quotasModule = shallowMount(QuotasModule, {
        store,
        localVue,
        mocks: {
          $t
        }
      });

      let result1 = quotasModule.vm.baseTwoRepresentation(0);
      let result2 = quotasModule.vm.baseTwoRepresentation(1395864371);
      let result3 = quotasModule.vm.baseTwoRepresentation(1363148);

      expect(result1).toBe(0);
      expect(result2).toBe(30);
      expect(result3).toBe(20);
    });

    it('should return correct unit if you give a base two', () => {
      const quotasModule = shallowMount(QuotasModule, {
        store,
        localVue,
        mocks: {
          $t
        }
      });

      let result1 = quotasModule.vm.getUnitFromBaseTwo(0);
      let result2 = quotasModule.vm.getUnitFromBaseTwo(32);
      let result3 = quotasModule.vm.getUnitFromBaseTwo(12);

      expect(result1).toBe('B');
      expect(result2).toBe('GB');
      expect(result3).toBe('Kb');
    });

    it('should divide the number passed by the base two provided', () => {
      const quotasModule = shallowMount(QuotasModule, {
        store,
        localVue,
        mocks: {
          $t
        }
      });

      let result1 = quotasModule.vm.divideBaseTwo(0, 0);
      let result2 = quotasModule.vm.divideBaseTwo(10, 0);
      let result3 = quotasModule.vm.divideBaseTwo(2097152, 20);
      let result4 = quotasModule.vm.divideBaseTwo(132070244352, 30);

      expect(result1).toBe(0);
      expect(result2).toBe(10);
      expect(result3).toBe(2);
      expect(result4).toBe(123);
    });
  });
});
