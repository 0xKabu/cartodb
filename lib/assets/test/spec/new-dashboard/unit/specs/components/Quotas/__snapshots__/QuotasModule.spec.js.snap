// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuotasModule.vue should render correct contents 1`] = `
<ul class="grid">
  <li class="vertical-space grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile">
    <quotacontainer-stub>
      <quotawidget-stub name="Limits.storage" amount="Limits.quota" availablecapacity="1024" usedcapacity="6.78125" unit="MB"></quotawidget-stub>
    </quotacontainer-stub>
  </li>
  <li class="vertical-space grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile">
    <quotacontainer-stub>
      <quotawidget-stub name="Limits.geocoding" amount="Limits.credits" availablecapacity="1200" usedcapacity="103"></quotawidget-stub>
    </quotacontainer-stub>
  </li>
  <li class="vertical-space grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile">
    <quotacontainer-stub>
      <quotawidget-stub name="Limits.isolines" amount="Limits.credits" availablecapacity="1332" usedcapacity="134" type="compact"></quotawidget-stub>
      <quotawidget-stub name="Limits.routing" amount="Limits.credits" availablecapacity="102" usedcapacity="54" type="compact"></quotawidget-stub>
    </quotacontainer-stub>
  </li>
</ul>
`;
